<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# **Retail Sales Dataset 2018-2022**
## **Lab 1. Loading the dataset**
Estimated time needed: **10** minutes

### **Context**
This dataset contains historical sales data. It was extracted from a global retail company. The data was transformed to protect the identity of the company.
The dataset is built from the initial dataset consisted of data collected in 5 years (period 2018-2022), indicating date, product and sales month. 
The goal is to load the dataset into your machine and familiarize yourself with the content using Python.

### **Dataset Attributes**
*   Date: year and month
*   SKU: unique code consisting of letters and numbers that identify each product
*   Group: group of related products which share some common attributes
*   Units Pkg: package weight (kg)
*   Avg Price Pkg: average price per package
*   Sales Pkg: total package sales per month

### **Target Field**
*   Turnover per month
### **Objectives**

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


## **Data Acquisition**
<p>There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://www.kaggle.com/datasets/tsmldata/retail-sales-dataset-2018-2022" target="_blank">https://www.kaggle.com/datasets/tsmldata/retail-sales-dataset-2018-2022</a></li>
    <li>Data type: csv</li>
    <li>License: <a href="https://creativecommons.org/publicdomain/zero/1.0/">https://creativecommons.org/publicdomain/zero/1.0/</a><br>
    You can copy, modify, distribute and perform the work, even for commercial purposes, all without asking permission.</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [1]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [1]:
# import pandas library
import pandas as pd
import numpy as np

## **Read Data**
<p>We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.

</p>


In [2]:
path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMSkillsNetwork-GPXX03NFEN/sales_1.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [3]:
# Read the online file by the URL provides above, and assign it to variable "df"
df = pd.read_csv(path)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [7]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


Date     SKU Group  Units_pkg  Avg_price_pkg  Sales_pkg
0  201801  SKU_01     A         13              5        879
1  201801  SKU_02     B          7              7          8
2  201801  SKU_03     B          7              6        197
3  201801  SKU_04     C          4              6          1
4  201801  SKU_05     C          6              8          1

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [6]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


Sometimes dataset may contain null values, so it's worth check this. We can do this this way:

In [7]:
print(df.isnull().sum())

Date             0
SKU              0
Group            0
Units_pkg        0
Avg_price_pkg    0
Sales_pkg        0
dtype: int64


As we can observe, there are not null values in our dataset. Everything is stored in a proper way.

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [8]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


## **Modify Headers**
<p>We can modify our headers. Let's do this.<br>
First, we print our headers by command <code>print(df.columns)</code></p>


In [9]:
# print headers list
print(df.columns)

Index(['Date', 'SKU', 'Group', 'Units_pkg', 'Avg_price_pkg', 'Sales_pkg'], dtype='object')


<p>We want to get rid of underscores. To do this we can use method <code>replace()</code></p>


In [10]:
df.columns = [col.replace("_", " ") for col in df.columns]

In [11]:
print(df.columns)

Index(['Date', 'SKU', 'Group', 'Units pkg', 'Avg price pkg', 'Sales pkg'], dtype='object')


<p>Looks better, but let's capiitalize every word in each header using method <code>title()</code> and leave column "SKU" as it is:


In [12]:
df.columns = [col.title() if col != "SKU" else col for col in df.columns]
        
print(df.columns)

Index(['Date', 'SKU', 'Group', 'Units Pkg', 'Avg Price Pkg', 'Sales Pkg'], dtype='object')


In [13]:
df.head(1500)

Date      SKU Group  Units Pkg  Avg Price Pkg  Sales Pkg
0     201801   SKU_01     A         13              5        879
1     201801   SKU_02     B          7              7          8
2     201801   SKU_03     B          7              6        197
3     201801   SKU_04     C          4              6          1
4     201801   SKU_05     C          6              8          1
...      ...      ...   ...        ...            ...        ...
1495  201911   SKU_97     J          4             10         56
1496  201911  SKU_117     J          4             10         36
1497  201911   SKU_93     J          4             10        221
1498  201911   SKU_84     J          5             11        110
1499  201911   SKU_85     J          5             11        139

[1500 rows x 6 columns]

Now, we have successfully read the raw dataset and modified headers. Looks much better.


# **Basic Insight of Dataset**
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


## **Data Types**
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [14]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

Date              int64
SKU              object
Group            object
Units Pkg         int64
Avg Price Pkg     int64
Sales Pkg         int64
dtype: object


As we can see, "SKU" and "Group" are <code>object</code> type, "Units Pkg", "Avg Price Pkg", "Date" and "Sales Pkg" are <code>int64</code> type.

These data types can be changed; we will learn how to accomplish this in a later module.

<p>
As shown above, it is clear to see that the data type of "Units Pkg", "Avg Price Pkg", "Sales Pkg" and "Date" are <code>int64</code>, "SKU" and "Group" are <code>object</code>.
</p><br>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


## **Describe**
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [15]:
df.describe()

Date    Units Pkg  Avg Price Pkg    Sales Pkg
count    4138.000000  4138.000000    4138.000000  4138.000000
mean   202015.618173    12.881344       9.138231   335.733446
std       142.616710    22.024843       3.551443   419.834537
min    201801.000000     0.000000       2.000000     0.000000
25%    201905.000000     5.000000       7.000000    65.000000
50%    202008.000000     5.000000       9.000000   180.000000
75%    202111.000000     7.000000      11.000000   436.750000
max    202212.000000   100.000000      26.000000  3364.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "Avg Price Pkg" has 4138 counts, the mean value of this column is 9.138231, the standard deviation is 3.551443, the minimum value is 2, 25th percentile is 7, 50th percentile is 9, 75th percentile is 11, and the maximum value is 26. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [16]:
# describe all the columns in "df" 
df.describe(include = "all")

Date     SKU Group    Units Pkg  Avg Price Pkg    Sales Pkg
count     4138.000000    4138  4138  4138.000000    4138.000000  4138.000000
unique            NaN     188    25          NaN            NaN          NaN
top               NaN  SKU_29     D          NaN            NaN          NaN
freq              NaN      60   592          NaN            NaN          NaN
mean    202015.618173     NaN   NaN    12.881344       9.138231   335.733446
std        142.616710     NaN   NaN    22.024843       3.551443   419.834537
min     201801.000000     NaN   NaN     0.000000       2.000000     0.000000
25%     201905.000000     NaN   NaN     5.000000       7.000000    65.000000
50%     202008.000000     NaN   NaN     5.000000       9.000000   180.000000
75%     202111.000000     NaN   NaN     7.000000      11.000000   436.750000
max     202212.000000     NaN   NaN   100.000000      26.000000  3364.000000

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'Date' and 'Units Pkg'.

</div>


In [17]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Date', 'Units Pkg']].describe()
```

</details>


## **Info**
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [18]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4138 entries, 0 to 4137
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Date           4138 non-null   int64 
 1   SKU            4138 non-null   object
 2   Group          4138 non-null   object
 3   Units Pkg      4138 non-null   int64 
 4   Avg Price Pkg  4138 non-null   int64 
 5   Sales Pkg      4138 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 194.1+ KB


As we can see, there aren't null values. Great!

## **Save Dataset**
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


In [19]:
df.to_csv("sales_1.csv", index=False)

We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


## **Read/Save Other Data Formats**

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


## **Excellent! You have just completed the Introduction Notebook!**


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
